[WOR-1544] #major Populate MDC for flight and its steps

okotsopoulos opened the pull request:
https://broadworkbench.atlassian.net/browse/WOR-1544

Centralizing work done in https://github.com/DataBiosphere/terra-workspace-manager/pull/1626, wrap flight thread execution so that flight threads are context-aware.

- Logs emitted during a flight thread now have the **context from the calling thread**, along with **flight-specific context** (flight ID and class).
- Logs emitted during the execution of a step additionally include **step-specific context** (step class, direction, and index). 

Handles the case where a flight thread spawns a child flight by intentionally cleaning any leftover step-specific context from the calling thread.

**In Action**

I published my Stairway branch locally, pulled it in as a WSM dependency, removed the existing WSM code which modifies the MDC, deployed it to my Bee, and hit an endpoint that triggers a stairway flight.

The resulting logs have their MDC populated as described in this PR: https://cloudlogging.app.goo.gl/T9vdJ86RXRoa117VA
- Calling thread's context over the entirety of the flight (that's where `jsonPayload.requestId` comes from)
- Flight-specific context over the entirety of the flight
- Step-specific context over the entirety of each step

A `jsonPayload` for a log emitted during a step:
<img width="1178" alt="Screenshot 2024-02-29 at 7 44 47 PM" src="https://github.com/DataBiosphere/stairway/assets/79769153/15e6be29-fcc9-4813-ba8c-27ec0162e1fd">

All of the logs for the request, with `flightId`, `flightStepNumber`, and `flightStepDirection` added to the summary line:
<img width="1696" alt="Screenshot 2024-02-29 at 7 46 57 PM" src="https://github.com/DataBiosphere/stairway/assets/79769153/cb07f652-4a9e-4002-ba6b-27fc7ececc13">

(As an aside, it's clear to me that these log lines can be substantially trimmed.  Do we need the full formatted summary anymore if we have this info in the MDC, and can pull it in the summary line?)

**How can my service get this new functionality?**

Once this PR is merged, I'll update the dependency in TCL, which is where most services get Stairway from.  At any rate, downstream services will need to update a dependency.

If downstream services presently modify their MDCs themselves, they will be able to remove that code in their service and let Stairway take care of it.

okotsopoulos committed:
Populate MDC for flight and its steps

Wrap flight thread execution so that flight threads are context-aware.
Logs emitted during a flight thread now have the context from the calling thread, along with flight-specific context (flight ID and class).
Logs emitted during the execution of a step additionally include step-specific context (step class, direction, and index).
Handles the case where a flight thread spawns a child flight by intentionally cleaning any leftover step-specific context from the calling thread.
Drive-by: now that we include flight ID and class on all flight logs in the MDC, we can trim down the dismal failure log line.

okotsopoulos committed:
Cannot call MDC.setContextMap with a null input

okotsopoulos committed:
Add unit tests, spotlessApply

In a follow-on PR, will automate spotlessCheck to gate PRs.

okotsopoulos commented on stairway/src/main/java/bio/terra/stairway/impl/FlightRunner.java:
```diff
@@ -144,11 +144,9 @@ private FlightStatus fly() throws InterruptedException {
-          flightContext.getFlightClassName(),
-          flightContext.getStepIndex(),
-          flightContext.getStepClassName());
+          "{} (index {}) experienced DISMAL FAILURE: non-retryable error on undo",
```
Drive-by: now that we include flight ID and class on all flight logs in the MDC, we can trim down the dismal failure log line.

I also reformatted to lead with the failing step class based off of Stairway retro feedback that all of the useful information on a log line is at the end. Do we like this better? Unfortunately we don't get step context in MDC for free at this point -- we're outside of step execution.

okotsopoulos committed:
Update README to highlight MDC additions

And some other formatting and documentation changes.

okotsopoulos committed:
MdcHelper, MdcHelperTest needn't be public

The MdcHelper methods are only meant to be called within the Stairway implementation package.

okotsopoulos commented on README.md:
```diff
@@ -93,6 +93,16 @@ Stairway is designed to provide atomic operations for one instance of one servic
+Stairway leverages the underlying logging system's mapped diagnostic context (MDC) if available.
+
+MDC manages contextual information on a per-thread basis: as Stairway submits a Flight for processing,
+it **passes along the MDC of the calling thread** so that context isn't lost.  It also further populates
```
Propagating the MDC from the calling thread makes me want to designate this as a minor version bump at minimum.  Default behavior in this repo is patch bumps.

Reviewers, WDYT?

cahrens replied:
I agree with a more significant version bump. Though with all the dependency upgrades underway, I was already thinking we would do a major version bump very soon. Perhaps just do a major bump with this change?

okotsopoulos replied:
Good idea. I'll make it a major one.

okotsopoulos replied:
Added #major to the PR title, which will get pulled into the commit message.

okotsopoulos commented on buildSrc/src/main/groovy/stairway.java-conventions.gradle:
```diff
@@ -16,8 +16,7 @@ dependencies {
     // For testing
-    testImplementation group: 'org.junit.jupiter', name: 'junit-jupiter-api', version: '5.9.0'
-    testImplementation group: 'org.junit.jupiter', name: 'junit-jupiter-engine', version: '5.9.0'
+    testImplementation group: 'org.junit.jupiter', name: 'junit-jupiter', version: '5.10.2'
```
Needed this expanded dependency for parameterized tests, and upgraded while in the neighborhood.

cahrens commented on stairway/src/main/java/bio/terra/stairway/impl/MdcHelper.java:
```diff
@@ -0,0 +1,76 @@
+    Map<String, String> contextMap = MDC.getCopyOfContextMap();
+    return () -> {
+      // Any leftover context on the thread will be fully overwritten:
+      MDC.setContextMap(Optional.ofNullable(contextMap).orElse(Map.of()));
```
I'm having trouble understanding this line. We are getting a copy of the context map at line 35, so why isn't line 38 effectively a no-op?

okotsopoulos replied:
Ah, great question.  Stairway spawns new threads for flight execution, so this Runnable within may be happening on a new thread without the prior context.

But this tells me that this code would be better organized if it's added to the Stairway thread pool implementation instead. I'll make that change.

okotsopoulos replied:
Hi @cahrens , I reorganized this code to live in `StairwayThreadPool` and feel like it makes more sense now, thanks for the original comment.  Would you mind taking another look?

cahrens commented on stairway/src/main/java/bio/terra/stairway/impl/MdcHelper.java:
```diff
@@ -0,0 +1,76 @@
+      // If the calling thread's context contains flight and step context from a parent flight, this
+      // will be overwritten below:
+      addFlightContextToMdc(flightContext);
+      removeStepContextFromMdc(flightContext);
```
I don't understand removing step context here.

okotsopoulos replied:
When I added flight and step context to the MDC in WSM's MdcHook, I didn't account for subflights that were kicked off within a parent flight's step.  The initial logs emitted by subflights before the first step errantly included step-specific context from the calling flight, interspersed with flight-specific context from the new subflight.  https://github.com/DataBiosphere/terra-workspace-manager/pull/1636

What changes would help make this clear?

cahrens replied:
Ah, I get it now.

cahrens commented on stairway/src/test/java/bio/terra/stairway/impl/MdcHelperTest.java:
```diff
@@ -0,0 +1,139 @@
+  }
+
+  static Stream<Map<String, String>> initialContext() {
+    return Stream.of(null, Map.of(), FOO_BAR);
```
Isn't this just returning a single map? If so, I don't see the point of making the tests parameterized.

okotsopoulos committed:
Code reorg: flight context-awareness moved to StairwayThreadPool

Because Stairway flights are submitted to the StairwayThreadPool for execution, it makes more sense to update the threadpool to wrap task submission to handle setting up and clearing flight context for the duration of the flight run on the child thread.
MdcHelper becomes MdcUtils, now a home for less involved utility methods for modifying the MDC.
Expanded tests accordingly: StairwayThreadPool did not previously have targeted unit tests.

okotsopoulos commented on stairway/src/main/java/bio/terra/stairway/impl/StairwayThreadPool.java:
```diff
@@ -29,10 +33,45 @@ int getQueuedFlights() {
-    super.beforeExecute(t, r);
     int active = activeTasks.incrementAndGet();
     logger.debug("before: " + active);
+    super.beforeExecute(t, r);
```
Reading the docstring on this method, it's recommended to call the super at the end to allow for potential nesting.

(I also read the docstring on `afterExecute`, and the super call at the beginning matches the recommendation.)

okotsopoulos committed:
Fix MdcUtilsTest.addFlightContextToMdc

okotsopoulos committed:
Enable Gradle scans on unit test runs

okotsopoulos committed:
Stabilize StairwayThreadPoolTests, flaky in CI

Removed checks on stairwayThreadPool.getActiveFlights, which is only used in debug logging.

cahrens commented on .github/workflows/build-and-test.yaml:
```diff
@@ -47,7 +47,7 @@ jobs:
           cache: 'gradle'
       - name: Run tests
-        run: ./gradlew test
+        run: ./gradlew test --scan
```
What does this do?

okotsopoulos replied:
This generates a Gradle build scan as part of test output, which makes debugging tests a lot easier.  Example from the most recent unit test run:

```
Publishing build scan...
https://gradle.com/s/gu4jckjm4mqzq
```

cahrens commented on stairway/src/main/java/bio/terra/stairway/impl/FlightRunner.java:
```diff
@@ -242,6 +240,7 @@ private StepResult stepWithRetry() throws InterruptedException, StairwayExecutio
     // Retry loop
     do {
       try {
+        MdcUtils.addStepContextToMdc(flightContext);
```
Ah, it makes a lot more sense here-- thanks!

cahrens commented on stairway/src/main/java/bio/terra/stairway/impl/MdcUtils.java:
```diff
@@ -0,0 +1,74 @@
+    if (context == null) {
+      MDC.clear();
+    } else {
+      MDC.setContextMap(context);
```
If there was already a context map, does this completely clear it (as opposed to merging)? 

I was just wondering if the following would be more robust/clearer:

```
MDC.clear();
if (context != null) {
  MDC.setContextMap(context);
}

okotsopoulos replied:
They are functionally the same per MDC's documentation, but I agree that your version is clearer.  Will update.

okotsopoulos committed:
PR feedback: rewrite MdcUtils.overwriteContext for clarity

And revisiting MdcUtilsTest with some naming changes for clarity.
